elastic/apm-server
# Universal Profiling: implement custom ILM strategy for K/V data

## Motivation/summary

In Universal Profiling beta we use both data streams and raw indices.
Elasticsearch does not provide a built-in ILM strategy to handle K/V data (discussing in https://github.com/elastic/elasticsearch/issues/85273), and we ought give users the ability to limit the storage used for profiling data.

We were able to provide ILM policies for data stream, but for raw indices we yet don't have nothing.
This PR provides basic ILM capabilities to rollover raw indices using a sliding window data lifecyle: the index age gates the rollover.

## Checklist


For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**? ✅ 
- Is its use being published in **telemetry** to enable product improvement? 🔴 
- Have system tests been added to avoid regression? ✅ 

## How to test these changes

Locally with runapm + fleetctl:
* start the docker compose stack and run the Elastic Agent container with `runapm`
* create an API key with the `admin` role
* fetch the package policy ID from Fleet for the APM input
* use `fleetctl` to configure profiling intake and the new ILM runners
```bash
go run main.go set-policy-config <PACKAGE_POLICY_ID>  apm-server.profiling.enabled=true apm-server.profiling.elasticsearch.api_key="<API_KEY>" apm-server.profiling.metrics.elasticsearch.hosts='["http://localhost:9200"]' apm-server.profiling.metrics.elasticsearch.api_key="<API_KEY>" apm-server.profiling.keyvalue_retention.execution_interval="10s" server.profiling.keyvalue_retention.age="10m"
```
* ingest data with a profiling host-agent connected to the apm-server intake
* observe the metrics `"beat.stats.apm_server.profiling.ilm.custom_ilm.execution.count"` and `"beat.stats.apm_server.profiling.ilm.custom_ilm.skipped_because_time_constraints.count"`, representing the counters of when indices rollover is taking place or skipped, respectively

## Related issues
